File: examples/.net/cases/tasks/task_create/task_create.agentic.case.cs
using AppProtocol.Example.DotNet.Core;
using AppProtocol.Example.DotNet.Core.Shared;

namespace AppProtocol.Example.DotNet.Cases.Tasks.TaskCreate;

public sealed class TaskCreateAgentic : BaseAgenticCase<TaskCreateInput, TaskCreateOutput>
{
    public TaskCreateAgentic(AgenticContext context)
        : base(context)
    {
    }

    protected override BaseDomainCase<TaskCreateInput, TaskCreateOutput> Domain() => new TaskCreateDomain();

    public override AgenticDiscovery Discovery() => new()
    {
        Name = DomainCaseName() ?? "task_create",
        Description = DomainDescription() ?? "Create a new task card for the board.",
        Category = "tasks",
        Tags = new[] { "tasks", "creation", "board" },
        Aliases = new[] { "create_task", "new_task_card", "add_board_task" },
        Capabilities = new[] { "task_creation" },
        Intents = new[] { "create a task", "add a task", "add work to the board" },
    };

    public override AgenticExecutionContext Context() => new()
    {
        RequiresAuth = false,
        RequiresTenant = false,
        Dependencies = new[] { "task_create.domain", "task_create.api" },
        Preconditions = new[] { "A non-empty title must be provided." },
        Constraints = new[]
        {
            "The caller must not provide id, status, createdAt, or updatedAt.",
            "Execution must delegate to the canonical API surface.",
            "Descriptions stay optional and should only be passed when the user supplied them.",
        },
        Notes = new[]
        {
            "New tasks always start in todo.",
            "The backend is the source of truth for identifiers and timestamps.",
        },
    };

    public override AgenticPrompt Prompt() => new()
    {
        Purpose = "Create a new task with a required title and an optional description.",
        WhenToUse = new[]
        {
            "When the user asks to create or add a new task card.",
            "When new work needs to be placed into the board backlog.",
        },
        WhenNotToUse = new[]
        {
            "When the user wants to inspect existing tasks.",
            "When the user wants to move an existing task between columns.",
        },
        Constraints = new[]
        {
            "Ask for a title if the user did not provide one.",
            "Do not invent backend-controlled fields.",
        },
        ReasoningHints = new[]
        {
            "Treat description as optional and pass it only when the user gave enough detail.",
            "Prefer concise titles because they are displayed directly on the board card.",
        },
        ExpectedOutcome = "A created task object with status todo and backend-generated identity fields.",
    };

    public override AgenticToolContract Tool()
    {
        var inputSchema = DomainInputSchema() ?? throw new InvalidOperationException("task_create.agentic requires domain schemas");
        var outputSchema = DomainOutputSchema() ?? throw new InvalidOperationException("task_create.agentic requires domain schemas");

        return new AgenticToolContract
        {
            Name = "task_create",
            Description = "Create a task through the canonical API execution flow.",
            InputSchema = inputSchema,
            OutputSchema = outputSchema,
            IsMutating = true,
            ExecuteAsync = async (input, ctx) =>
            {
                var handler = ResolveApiHandler(ctx);
                var responseObject = await handler(Materialize<TaskCreateInput>(input));
                var response = Materialize<ApiResponse<TaskCreateOutput>>(responseObject);
                if (!response.Success || response.Data is null)
                {
                    throw AppCaseErrors.ToAppCaseError(response.Error, "task_create API failed");
                }

                return response.Data;
            },
        };
    }

    public override AgenticMcpContract Mcp() => new()
    {
        Enabled = true,
        Name = "task_create",
        Title = "Create Task",
        Description = "Create a board task through the canonical APP task_create API flow.",
        Metadata = new Dictionary<string, object?>
        {
            ["category"] = "tasks",
            ["mutating"] = true,
        },
    };

    public override AgenticRagContract Rag() => new()
    {
        Topics = new[] { "task_management", "board_backlog", "task_creation" },
        Resources = new[]
        {
            new AgenticKnowledgeResource
            {
                Kind = "case",
                Ref = "tasks/task_create",
                Description = "Canonical task creation capability for new board work.",
            },
            new AgenticKnowledgeResource
            {
                Kind = "case",
                Ref = "tasks/task_list",
                Description = "Board grounding capability for inspecting current tasks before adding related work.",
            },
        },
        Hints = new[]
        {
            "Prefer the canonical backlog language used by the board.",
            "Keep task titles concise because they render directly on cards.",
        },
        Scope = "project",
        Mode = "recommended",
    };

    public override AgenticPolicy Policy() => new()
    {
        RequireConfirmation = false,
        RiskLevel = "low",
        ExecutionMode = "direct-execution",
        Limits = new[] { "Use only for explicit task-creation intent." },
    };

    public override async Task TestAsync()
    {
        ValidateDefinition();
        var definition = Definition();
        if (definition.Tool.IsMutating != true)
        {
            throw new InvalidOperationException("test: task_create agentic must be mutating");
        }

        var example = Examples()?.FirstOrDefault(item => item.Name == "title_only")
            ?? throw new InvalidOperationException("test: task_create example must exist");

        var mockCtx = new AgenticContext
        {
            CorrelationId = "task-create-agentic-test",
            Logger = Ctx.Logger,
            Cases = new Dictionary<string, IDictionary<string, IDictionary<string, object>>>
            {
                ["tasks"] = new Dictionary<string, IDictionary<string, object>>
                {
                    ["task_create"] = new Dictionary<string, object>
                    {
                        ["api"] = new Func<object?, Task<object?>>(_ => Task.FromResult<object?>(
                            ApiResponse<TaskCreateOutput>.Ok(Materialize<TaskCreateOutput>(example.Output)))),
                    },
                },
            },
        };

        var result = Materialize<TaskCreateOutput>(await Tool().ExecuteAsync(example.Input, mockCtx));
        if (result.Task.Status != "todo")
        {
            throw new InvalidOperationException("test: task_create tool must return a todo task");
        }

        try
        {
            await Tool().ExecuteAsync(example.Input, new AgenticContext
            {
                CorrelationId = "task-create-agentic-failure-test",
                Logger = Ctx.Logger,
                Cases = new Dictionary<string, IDictionary<string, IDictionary<string, object>>>
                {
                    ["tasks"] = new Dictionary<string, IDictionary<string, object>>
                    {
                        ["task_create"] = new Dictionary<string, object>
                        {
                            ["api"] = new Func<object?, Task<object?>>(_ => Task.FromResult<object?>(
                                ApiResponse<TaskCreateOutput>.Failure(new AppError("VALIDATION_FAILED", "title must not be empty"), 400))),
                        },
                    },
                },
            });
        }
        catch (AppCaseError error) when (error.Code == "VALIDATION_FAILED")
        {
            return;
        }

        throw new InvalidOperationException("test: task_create must propagate AppCaseError failures");
    }

    private static Func<object?, Task<object?>> ResolveApiHandler(AgenticContext context)
    {
        if (context.Cases is null ||
            !context.Cases.TryGetValue("tasks", out var taskCases) ||
            !taskCases.TryGetValue("task_create", out var surfaces) ||
            !surfaces.TryGetValue("api", out var apiSurface) ||
            apiSurface is not Func<object?, Task<object?>> handler)
        {
            throw new AppCaseError("NOT_FOUND", "task_create API surface is not available in ctx.cases");
        }

        return handler;
    }
}

File: examples/python/cases/tasks/task_create/task_create_agentic_case.py
# ==========================================================================
# APP v1.1.1
# cases/tasks/task_create/task_create_agentic_case.py
# --------------------------------------------------------------------------
# Agentic surface for the task_create Case.
#
# Responsibility:
# - make the task_create capability understandable by agents
# - expose discovery, context, prompt, tool, MCP, RAG, and policy
# - delegate real execution to the canonical API surface
# - never reimplement the business logic
# ==========================================================================

from __future__ import annotations

from typing import Any, Literal, TypedDict

from core.agentic_case import (
    AgenticContext,
    AgenticDiscovery,
    AgenticExample,
    AgenticExecutionContext,
    AgenticMcpContract,
    AgenticPolicy,
    AgenticPrompt,
    AgenticRagContract,
    AgenticToolContract,
    BaseAgenticCase,
    RagResource,
)
from core.shared.app_structural_contracts import AppCaseError, to_app_case_error

from cases.tasks.task_create.task_create_domain_case import (
    Task,
    TaskCreateDomain,
    TaskCreateInput,
    TaskCreateOutput,
    TaskStatus,
)

# Re-export shared types
__all__ = [
    "TaskCreateAgentic",
    "Task",
    "TaskCreateInput",
    "TaskCreateOutput",
    "TaskStatus",
]


# ==========================================================================
# TaskCreateAgentic
# ==========================================================================

class TaskCreateAgentic(BaseAgenticCase[TaskCreateInput, TaskCreateOutput]):
    """Agentic surface for the task_create Case."""

    def domain(self) -> TaskCreateDomain:
        """Return the local domain instance for schema and example derivation."""
        return TaskCreateDomain()

    def discovery(self) -> AgenticDiscovery:
        return AgenticDiscovery(
            name="task_create",
            description=(
                self.domain_description()
                or "Create a new task card for the board."
            ),
            category="tasks",
            tags=["task", "create"],
            capabilities=["create_task"],
            intents=["create a new task"],
        )

    def context(self) -> AgenticExecutionContext:
        return AgenticExecutionContext(
            requires_auth=False,
            requires_tenant=False,
            dependencies=[],
            preconditions=[],
            constraints=[
                "A non-empty title must be provided.",
                "The caller must not provide id, status, createdAt, or updatedAt.",
            ],
            notes=[
                "New tasks always start in todo.",
                "The backend is the source of truth for identifiers and timestamps.",
            ],
        )

    def prompt(self) -> AgenticPrompt:
        return AgenticPrompt(
            purpose="Create a new task with a required title.",
            when_to_use=[
                "When the user asks to create or add a new task card.",
                "When new work needs to be placed into the board backlog.",
            ],
            when_not_to_use=[
                "When the user wants to inspect existing tasks.",
                "When the user wants to move an existing task between columns.",
            ],
            constraints=[
                "Ask for a title if the user did not provide one.",
                "Do not invent backend-controlled fields.",
            ],
            expected_outcome=(
                "A created task object with status todo and backend-generated identity fields."
            ),
        )

    def tool(self) -> AgenticToolContract[TaskCreateInput, TaskCreateOutput]:
        """
        Tool contract for task_create.

        execute() delegates to ctx.cases["tasks"]["task_create"]["api"].handler(input).
        """
        input_schema = self.domain_input_schema()
        output_schema = self.domain_output_schema()

        if input_schema is None or output_schema is None:
            raise AppCaseError(
                "INTERNAL",
                "task_create.agentic requires domain schemas",
            )

        def execute(input: TaskCreateInput, ctx: AgenticContext) -> TaskCreateOutput:
            cases: Any = ctx.get("cases", {})
            api_handler = (
                cases.get("tasks", {})
                .get("task_create", {})
                .get("api", {})
                .get("handler")
            )

            if not callable(api_handler):
                raise AppCaseError(
                    "INTERNAL",
                    "task_create.agentic: ctx.cases.tasks.task_create.api.handler is not available",
                )

            result = api_handler(input)

            if not result.success or result.data is None:
                raise to_app_case_error(result.error, "task_create API failed")

            return result.data

        return AgenticToolContract(
            name="task_create",
            description=self.domain_description() or "Create a task through the canonical API execution flow.",
            input_schema=input_schema,
            output_schema=output_schema,
            is_mutating=True,
            requires_confirmation=False,
            execute=execute,
        )

    def mcp(self) -> AgenticMcpContract:
        return AgenticMcpContract(
            enabled=True,
            name="task_create",
            title="Create Task",
        )

    def rag(self) -> AgenticRagContract:
        return AgenticRagContract(
            topics=["task_management"],
            resources=[
                RagResource(
                    kind="case",
                    ref="tasks/task_create",
                    description="Canonical task creation capability for new board work.",
                ),
            ],
            scope="case-local",
            mode="optional",
        )

    def policy(self) -> AgenticPolicy:
        return AgenticPolicy(
            require_confirmation=False,
            risk_level="low",
            execution_mode="direct-execution",
        )

    def examples(self) -> list[AgenticExample[TaskCreateInput, TaskCreateOutput]]:
        """Derive examples from the domain surface."""
        return self.domain_examples() or []

    def test(self) -> None:
        """Self-contained agentic surface test."""

        # --- Structural validation ---
        self.validate_definition()

        definition = self.definition()

        if not definition.tool.is_mutating:
            raise AssertionError("test: task_create agentic must be mutating")

        if definition.policy is None or definition.policy.execution_mode != "direct-execution":
            raise AssertionError("test: task_create should default to direct execution")

        if definition.tool.name != "task_create":
            raise AssertionError("test: tool.name must be task_create")

        if not definition.tool.requires_confirmation:
            pass  # Correct — task_create does not require confirmation

        # --- Verify tool execute delegates correctly via mock context ---
        example = next(
            (e for e in self.examples() if e.name == "title_only"),
            None,
        )
        if example is None:
            raise AssertionError("test: title_only example must exist")

        mock_ctx: AgenticContext = {
            "correlation_id": "task-create-agentic-test",
            "logger": self.ctx["logger"],
            "cases": {
                "tasks": {
                    "task_create": {
                        "api": {
                            "handler": lambda _input: _make_success_response(example.output),
                        },
                    },
                },
            },
        }

        result = self.tool().execute(example.input, mock_ctx)
        if result["task"]["status"] != "todo":
            raise AssertionError("test: tool execute must return a todo task")

        # --- Verify errors propagate correctly ---
        propagated_error: AppCaseError | None = None
        try:
            self.tool().execute(
                example.input,
                {
                    "correlation_id": "task-create-agentic-failure-test",
                    "logger": self.ctx["logger"],
                    "cases": {
                        "tasks": {
                            "task_create": {
                                "api": {
                                    "handler": lambda _input: _make_failure_response(
                                        "VALIDATION_FAILED", "title must not be empty"
                                    ),
                                },
                            },
                        },
                    },
                },
            )
        except AppCaseError as exc:
            propagated_error = exc

        if propagated_error is None:
            raise AssertionError("test: task_create must propagate AppCaseError failures")

        if propagated_error.code != "VALIDATION_FAILED":
            raise AssertionError(
                f"test: task_create must preserve error codes, got {propagated_error.code}"
            )


# ==========================================================================
# Test helpers (module-private)
# ==========================================================================

class _FakeApiResponse:
    """Minimal ApiResponse-like object for test mocks."""

    def __init__(
        self,
        success: bool,
        data: TaskCreateOutput | None = None,
        error_code: str | None = None,
        error_message: str | None = None,
    ) -> None:
        self.success = success
        self.data = data
        from core.shared.app_structural_contracts import AppError
        self.error = (
            AppError(code=error_code or "INTERNAL", message=error_message or "")
            if not success
            else None
        )


def _make_success_response(output: TaskCreateOutput) -> _FakeApiResponse:
    return _FakeApiResponse(success=True, data=output)


def _make_failure_response(code: str, message: str) -> _FakeApiResponse:
    return _FakeApiResponse(success=False, error_code=code, error_message=message)
